forch/endpoint_handler.py

```diff
@@ -13,54 +13,100 @@
 import logging
 import subprocess
 import time
+import threading
+import yaml
 import grpc
 
 from forch.utils import get_logger
-
-try:
-    import daq.proto.session_server_pb2_grpc as server_grpc
-    from daq.proto.session_server_pb2 import SessionParams, SessionProgress
-    SessionServerServicer = server_grpc.SessionServerServicer
-    PROTO_LOADED = True
-except ImportError:
-    PROTO_LOADED = False
-
-    class SessionServerServicer:
-        """Dummy class for weak import"""
+import forch.proto.endpoint_server_pb2_grpc as server_grpc
+from forch.proto.endpoint_server_pb2 import Endpoint
+from forch.proto.shared_constants_pb2 import Empty
 
 DEFAULT_SERVER_PORT = 50111
 DEFAULT_BIND_ADDRESS = '0.0.0.0'
 DEFAULT_VXLAN_PORT = 4789
-DEFAULT_VXLAN_VNI = 0
-
+BASE_T1SW_PORT = 29
 VXLAN_CMD_FMT = 'ip link add %s type vxlan id %s remote %s dstport %s srcport %s %s nolearning'
-
+T1_SW1 = 'nz-kiwi-t1sw1'
 CONNECT_TIMEOUT_SEC = 60
 
 
 class EndpointHandler:
     """Class to handle endpoint updates"""
 
-    def __init__(self, target_ip):
+    def __init__(self, target_ip, structural_config_file):
         self._logger = get_logger('endpproxy')
+        self._lock = threading.RLock()
+        self._mac_t1sw1_port = {}
+        self._freed_t1sw_ports = set()
+        self._next_t1sw_port = BASE_T1SW_PORT
+        self._structural_config_file = structural_config_file
         server_port = DEFAULT_SERVER_PORT
         address = f'{target_ip}:{server_port}'
-        self._logger.info('Proxy requests to %s, proto %s', address, PROTO_LOADED)
+        self._logger.info('Proxy requests to %s', address)
         channel = grpc.insecure_channel(address)
-        self._stub = server_grpc.SessionServerStub(channel)
+        self._stub = server_grpc.EndpointServerStub(channel)
         grpc.channel_ready_future(channel).result(timeout=CONNECT_TIMEOUT_SEC)
 
-    def process_endpoint(self, endpoint):
+    def _allocate_t1sw_port(self, mac):
+        with self._lock:
+            if mac in self._mac_t1sw1_port:
+                return self._mac_t1sw1_port[mac]
+            if self._freed_t1sw_ports:
+                t1sw_port = max(self._freed_t1sw_ports)
+                self._freed_t1sw_ports.remove(t1sw_port)
+            else:
+                t1sw_port = self._next_t1sw_port
+                self._next_t1sw_port += 1
+            self._mac_t1sw1_port[mac] = t1sw_port
+            with open(self._structural_config_file, 'r') as file:
+                structural_config = yaml.safe_load(file)
+                structural_config['dps'][T1_SW1]['interfaces'][t1sw_port] = {
+                    'description': 'TAP',
```

Comment:
nit: _ideally_ (and not necessary to change), this info (description and tagged_vlans) should be copied from the already present TAP port that's there. But, that might be unnecessary work given the limited lifespan of this code.

At the very least please extract to function (the port dict info), so that it's conceptually separated and then add a TODO at the function level of how the information is determined.